lib/elastic_apm/timed_worker.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+module ElasticAPM
+  # @api private
+  class TimedWorker
+    include Log
+
+    SLEEP_INTERVAL = 0.1
+
+    # @api private
+    class StopMsg
+    end
+
+    # @api private
+    class ErrorMsg
+      def initialize(error)
+        @error = error
+      end
+
+      attr_reader :error
+    end
+
+    def initialize(config, messages, pending_transactions, adapter)
+      @config = config
+      @messages = messages
+      @pending_transactions = pending_transactions
+      @adapter = adapter
+
+      @last_sent_transactions = Time.now.utc
+
+      @serializers = Struct.new(:transactions, :errors).new(
+        Serializers::Transactions.new(config),
+        Serializers::Errors.new(config)
+      )
+    end
+
+    attr_reader :config, :messages, :pending_transactions
+
+    def run_forever
+      loop do
+        run_once
+        sleep SLEEP_INTERVAL
+      end
+    end
+
+    def run_once
+      collect_and_send_transactions if should_flush_transactions?
+      process_messages
+    end
+
+    private
+
+    def process_messages
+      while (msg = messages.pop(true))
+        case msg
+        when ErrorMsg
+          post_error msg
+        when StopMsg
+          # empty collected transactions before exiting
+          collect_and_send_transactions
+
+          Thread.exit
+        end
+      end
+    rescue ThreadError
+      # queue empty
+    end
+
+    def post_error(msg)
+      payload = @serializers.errors.build_all([msg.error])
+      @adapter.post('/v1/errors', payload)
+    end
+
+    def collect_and_send_transactions
+      return if pending_transactions.empty?
+
+      transactions = collect_batched_transactions
+
+      payload = @serializers.transactions.build_all(transactions)
+
+      begin
+        @adapter.post('/v1/transactions', payload)
+      rescue ::Exception => e
+        fatal 'Failed posting: %s', e.inspect
+        debug e.backtrace.join("\n")
+        nil
+      end
+    end
+
+    def collect_batched_transactions
+      batch = []
+
+      begin
+        while (transaction = pending_transactions.pop(true))
```

Comment:
Consuming the the `transactions` here could become an issue, when events are enqueued faster than they can be processed. If new events get pushed to the queue fast enough, the condition `while queue.pop(true)` is always fulfilled. This would eventually lead to only a few requests to the server with large payloads. 

It might be an edge case, but if there are tons of requests to an endpoint I think this could happen. You probably need to additionally set a limit when consuming from the queue so the number of transactions that are sent within 1 request cannot grow endlessly.

Reply:
Oh yeah, you are right. What do you suggest? Something like a limit to how many transaction can be in a batch? 
`while pop && batch.length < MAX_BATCH_SIZE` ?

Reply:
Yes adding a `batch.length` check. How about you use the configured `max_queue_size` for an upper limit?